libvmi/driver/kvm/kvm.c

```diff
@@ -931,3 +1308,327 @@ kvm_resume_vm(
+        return VMI_FAILURE;
+    }
+
+    // pop event from queue
+    if (kvmi_pop_event(kvm->kvmi_dom, &event)) {
+        errprint("%s: kvmi_pop_event failed: %s\n", __func__, strerror(errno));
+        return VMI_FAILURE;
+    }
+
+    // handle event
+    ev_reason = event->event.common.event;
+#ifdef ENABLE_SAFETY_CHECKS
+    if ( ev_reason >= KVMI_NUM_EVENTS || !kvm->process_event[ev_reason] ) {
+        errprint("Undefined handler for %u event reason\n", ev_reason);
+        return VMI_FAILURE;
+    }
+#endif
+    // call handler
+    if (VMI_FAILURE == kvm->process_event[ev_reason](vmi, event))
+        goto error_exit;
+
+    return VMI_SUCCESS;
+error_exit:
+    if (event)
+        free(event);
+    return VMI_FAILURE;
+}
+
+
+status_t kvm_set_reg_access(
+    vmi_instance_t vmi,
+    reg_event_t* event)
+{
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!vmi || !event)
+        return VMI_FAILURE;
+#endif
+    unsigned int event_flags = 0;
+    unsigned int kvmi_reg = 0;
+    bool enable = false;
+
+    kvm_instance_t *kvm = kvm_get_instance(vmi);
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!kvm || !kvm->kvmi_dom)
+        return VMI_FAILURE;
+#endif
+
+    // check reg
+    switch (event->reg) {
+        case CR0:
+            event_flags |= KVMI_EVENT_CR_FLAG;
+            kvmi_reg = 0;
+            break;
+        case CR3:
+            event_flags |= KVMI_EVENT_CR_FLAG;
+            kvmi_reg = 3;
+            break;
+        case CR4:
+            event_flags |= KVMI_EVENT_CR_FLAG;
+            kvmi_reg = 4;
+            break;
+        case MSR_ALL:
+            event_flags |= KVMI_EVENT_MSR_FLAG;
+            break;
+        case MSR_ANY:
+            event_flags |= KVMI_EVENT_MSR_FLAG;
+            kvmi_reg = event->msr;
+            break;
+        case MSR_FLAGS ... MSR_TSC_AUX:
+        case MSR_STAR ... MSR_HYPERVISOR:
+            errprint("%s error: use MSR_ANY type for specific MSR event registration\n", __FUNCTION__);
+            return VMI_FAILURE;
+        default:
+            errprint("%s: unhandled register %" PRIu64"\n", __func__, event->reg);
+            return VMI_FAILURE;
+    }
+
+    // check access type
+    switch (event->in_access) {
+        case VMI_REGACCESS_N:
+            enable = false;
+            break;
+        case VMI_REGACCESS_W:
+            enable = true;
+            break;
+        case VMI_REGACCESS_R:
+        case VMI_REGACCESS_RW:
+            errprint("Register read events are unavailable in KVM.\n");
+            return VMI_FAILURE;
+        default:
+            errprint("Unknown register access mode: %d\n", event->in_access);
+            return VMI_FAILURE;
+    }
+
+    // handle MSR_ALL here instead of the switch before
+    if (event->reg != MSR_ALL) {
+        // enable event monitoring for all vcpus
+        for (unsigned int i = 0; i < vmi->num_vcpus; i++) {
+            if (kvmi_control_events(kvm->kvmi_dom, i, event_flags)) {
```

Comment:
@adlazar: does enabling the interception of KVMi events, without generating this type of events, has a slight performance impact, or close to none ?

The idea is that I would like to move enabling/disabling  event interception at driver init, so I could enable all of them at once, and disable them at teardown.

Otherwise, with the current implementation, we have to keep track of what `reg_access` has been enabled for example, or the following use case won't work:
- enable an event to intercept `CR3`
- enable an event to intercept `CR0`
- disable `CR0` event -> `CR3 is disabled too`..

Reply:
For MSR and CR events, you can enable them with `kvmi_control_events()` without worrying about the the performance impact, because the interception _does something_ when you enable a specific MSR/CR (`kvmi_control_msr`/`kvmi_control_cr`).

But, if you enable `KVMI_EVENT_BREAKPOINT` and `KVMI_EVENT_DESCRIPTOR`, you might get a lot of events when the guest starts. For the second one we might filter out the read events in the future (intercepting only the write accesses).